### PR TITLE
[TOPIC-BLE-LLCP] Bluetooth: controller: add Kconfig option for new llcp

### PR DESCRIPTION
**Pull request by kruithofa**
Added a new option for selecting the refactored LLCP.
Renamed the existing option to CONFIG_BT_LL_SW_SPLIT_LEGACY to minimise
the changes required in existing code due to conditional compiles

In the CMakeLists.txt file the correct implementation of LLCP is chosen
by selecting the corresping source-files

Signed-off-by: Andries Kruithof <Andries.Kruithof@nordicsemi.no>

### CODE REVIEW

**Comment by joerchan** [comment]
What is the reasoning behind putting the new implementation behind a Kconfig option?

I would prefer switching unconditionally.

**Comment by kruithofa** [comment]
> What is the reasoning behind putting the new implementation behind a Kconfig option?
> 
> I would prefer switching unconditionally.

Same reasoning as when the split architecture was introduced originally. 
Typically when more people use the new implementation this will uncover bugs, which may be time-consuming to fix.
Therefor we do not want to force all users to start using the new architecture when we release it for the first time.
The next step will be to make the new implementation the default (increasing the userbase further), and the final step will then be to remove the legacy implementation

**Comment by joerchan** [comment]
> Same reasoning as when the split architecture was introduced originally.
Typically when more people use the new implementation this will uncover bugs, which may be time-consuming to fix.

Yeah, but for a long time there we had to fix all bugs for both implementations, as well as some of the new features. I don't think we are saving ourselves from additional work with this option, in fact I think we would create more work.

**Comment by kruithofa** [comment]
> > Same reasoning as when the split architecture was introduced originally.
> > Typically when more people use the new implementation this will uncover bugs, which may be time-consuming to fix.
> 
> Yeah, but for a long time there we had to fix all bugs for both implementations, as well as some of the new features. I don't think we are saving ourselves from additional work with this option, in fact I think we would create more work.

Yes, maintaining two versions creates more work, no doubt about that. Still IMO we should not force people to use the new implementation from day 1, but this is something we will re-evaluate when pulling the topic-branch into master.

**Comment by joerchan** [comment]
@kruithofa Ok, fair enough. I have expressed my concerns atleast. :)

**Review by thoh-ot — CHANGES_REQUESTED** [review]
Didn't we agree on a meeting that we would go for a solution with only one KConfig setting?

File: subsys/bluetooth/controller/Kconfig [hunk 1]
```diff
@@ -68,18 +71,28 @@ choice BT_LL_CHOICE
 	help
 	  Select the Bluetooth Link Layer to compile.
 
-config BT_LL_SW_SPLIT
+config BT_LL_SW_LLCP_LEGACY
 	bool "Software-based BLE Link Layer"
+	select BT_LL_SW
 	select BT_RECV_IS_RX_THREAD
```
Comment (thoh-ot) [inline subsys/bluetooth/controller/Kconfig:77]: Not a fan of this construct. I would keep BT_LL_SW as is and then under the  "if BT_LL_SW" below i would add a BT_CTRL_LLCP_LEGACY config that is default y.

Comment (kruithofa) [reply subsys/bluetooth/controller/Kconfig:73]: Implemented as suggested

File: doc/guides/bluetooth/bluetooth-arch.rst [hunk 2]
```diff
@@ -110,7 +110,7 @@ BLE-enabled builds that can be produced from the Zephyr project codebase:
   * :option:`CONFIG_BT_HCI` ``=y``
   * :option:`CONFIG_BT_HCI_RAW` ``=y``
   * :option:`CONFIG_BT_CTLR` ``=y``
-  * :option:`CONFIG_BT_LL_SW_SPLIT` ``=y`` (if using the open source Link Layer)
+  * :option:`CONFIG_BT_LL_SW` ``=y`` (if using the open source Link Layer)
```
Comment (joerchan) [inline doc/guides/bluetooth/bluetooth-arch.rst:113]: Are you sure you want to do this change?
This is likely to exist in many prj.conf in existing applications and in our downstream repositories. You are creating a big clean-up job for yourself.

@rugeGerritsen Any opinions on maintenance for this rename?

Comment (rugeGerritsen) [reply doc/guides/bluetooth/bluetooth-arch.rst:113]: If there no important reason to do the change, I would leave it as is. I did a similar change renaming the nRF BLE Controller to SoftDevice controller. It required updates in multiple repos. 

If you decide to rename the configuration, you should add least keep the old one for deprecation purposes

**Review by joerchan — DISMISSED** [review]
I don't think it is wise to rename CONFIG_BT_LL_SW.

Atleast not as part of another commit doing something else.

File: subsys/bluetooth/controller/Kconfig [hunk 3]
```diff
@@ -495,6 +495,15 @@ config BT_CTLR_ADVANCED_FEATURES
 menu "Advanced features"
 	visible if BT_CTLR_ADVANCED_FEATURES
 
+config BT_LL_SW_LLCP_LEGACY
+	bool "Software-based BLE Link Layer"
```
Comment (erbr-ot) [inline subsys/bluetooth/controller/Kconfig:499]: Shouldn't the caption be something like "Use Legacy LLCP implementation" as well as an updated Help section?

File: doc/guides/bluetooth/bluetooth-arch.rst (hunk 2 above)
Comment (kruithofa) [reply doc/guides/bluetooth/bluetooth-arch.rst]: I understand the arguments here, and will not rename the configuration

**Comment by kruithofa** [comment]
> I don't think it is wise to rename CONFIG_BT_LL_SW.
> 
> Atleast not as part of another commit doing something else.

Reversed the name-change

File: subsys/bluetooth/controller/Kconfig.ll_sw_split [hunk 4]
```diff
@@ -107,6 +107,14 @@ config BT_CTLR_ADVANCED_FEATURES
 menu "Advanced features"
 	visible if BT_CTLR_ADVANCED_FEATURES
 
+config BT_LL_SW_SPLIT_LLCP_LEGACY
+	bool "Legacy software-based BLE Link Layer"
+	default y
+	select BT_RECV_IS_RX_THREAD
+	select ENTROPY_GENERATOR
```
Comment (thoh-ot) [inline subsys/bluetooth/controller/Kconfig.ll_sw_split:114]: Not needed for this config

**Review by thoh-ot — CHANGES_REQUESTED** [review]

**Review by thoh-ot — APPROVED** [review]

**Pull request merged**